src/memray/_memray.pyx

```diff
@@ -602,27 +614,104 @@ cdef class FileReader:
-            PythonAllocatorType.PYTHON_ALLOCATOR_PYMALLOC: "pymalloc",
-            PythonAllocatorType.PYTHON_ALLOCATOR_PYMALLOC_DEBUG: "pymalloc debug",
-            PythonAllocatorType.PYTHON_ALLOCATOR_MALLOC: "malloc",
-            PythonAllocatorType.PYTHON_ALLOCATOR_OTHER: "unknown",
-        }
-        python_allocator = allocator_id_to_name[self._header["python_allocator"]]
-        return Metadata(start_time=millis_to_dt(stats["start_time"]),
-                        end_time=millis_to_dt(stats["end_time"]),
-                        total_allocations=stats["n_allocations"],
-                        total_frames=stats["n_frames"],
-                        peak_memory=self._high_watermark.peak_memory,
-                        command_line=self._header["command_line"],
-                        pid=self._header["pid"],
-                        python_allocator=python_allocator,
-                        has_native_traces=self._header["native_traces"])
+    # Ignore the n_allocations in the header, use our observed value.
+    header["stats"]["n_allocations"] = progress_indicator.num_processed
+
+    # NOTE: We need to cast away the constness of some references below,
+    #       because Cython's iterator handling is not const-correct.
+
+    # Convert allocation counts by allocator/by size to Python dicts.
+    allocation_count_by_allocator = {
+        AllocatorType(<int>id_and_count.first).name: id_and_count.second
+        for id_and_count in <unordered_map[Allocator, uint64_t]&>(
+            aggregator.allocationCountByAllocator()
+        )
+    }
+    allocation_count_by_size = {
+        size_and_count.first: size_and_count.second
+        for size_and_count in <unordered_map[size_t, uint64_t]&>(
+            aggregator.allocationCountBySize()
+        )
+    }
+
+    # We've aggregated sizes and counts by stack, but we want to aggregate them
+    # by source location instead. Ideally we wouldn't do this in Python, but
+    # there's currently no other way to look up a frame by ID.
+    size_and_count_by_location = collections.defaultdict(lambda: [0, 0])
+    for location_key_and_size_and_count in <AllocationStatsAggregator.SizeAndCountByStack&>(
+        aggregator.sizeAndCountByStack()
+    ):
+        top_of_stack = reader.Py_GetStackFrame(
+            location_key_and_size_and_count.first.python_frame_id, 1
+        )
+        if not top_of_stack:
+            top_of_stack = [("<unknown>", "<unknown>", 0)]
+        size_and_count = size_and_count_by_location[top_of_stack[0]]
+        size_and_count[0] += location_key_and_size_and_count.second.first
+        size_and_count[1] += location_key_and_size_and_count.second.second
+
+    # Now we've got sizes and counts by source location. Find the top locations.
+    top_locations_by_size = [
+        (loc, size_and_count[0])
+        for loc, size_and_count in size_and_count_by_location.items()
+    ]
+    top_locations_by_size.sort(key=lambda pair: pair[1], reverse=True)
```

Comment:
Nit: Using `operator.itemgetter` may be better than the lambda

Reply:
I rewrote this to use `heapq.nlargest` and avoided the need for `key` entirely.